packages/circuits/test/proveGraffitiPreimage.test.ts

```diff
@@ -0,0 +1,147 @@
+import * as path from 'path'
+import { expect } from 'chai'
+import { genRandomSalt, ZkIdentity } from '@unirep/crypto'
+import { genEpochKey, Reputation } from '@unirep/core'
+import * as crypto from '@unirep/crypto'
+import * as snarkjs from 'snarkjs'
+import {
+    USER_STATE_TREE_DEPTH,
+    GLOBAL_STATE_TREE_DEPTH,
+} from '@unirep/circuits'
+
+const circuitName = 'proveGraffitiPreimage'
+
+const genCircuitInput = (
+    id: crypto.ZkIdentity,
+    epoch: number,
+    reputationRecords,
+    attesterId
+) => {
+    if (reputationRecords[attesterId] === undefined) {
+        reputationRecords[attesterId] = new Reputation(
+            BigInt(0),
+            BigInt(0),
+            BigInt(0),
+            BigInt(0)
+        )
+    }
+
+    // User state tree
+    const userStateTree = new crypto.SparseMerkleTree(
+        USER_STATE_TREE_DEPTH,
+        crypto.hash5([BigInt(0), BigInt(0), BigInt(0), BigInt(0), BigInt(0)])
+    )
+    for (const attester of Object.keys(reputationRecords)) {
+        userStateTree.update(
+            BigInt(attester),
+            reputationRecords[attester].hash()
+        )
+    }
+    const userStateRoot = userStateTree.root
+    const USTPathElements = userStateTree.createProof(BigInt(attesterId))
+
+    // Global state tree
+    const GSTree = new crypto.IncrementalMerkleTree(GLOBAL_STATE_TREE_DEPTH)
+    const commitment = id.genIdentityCommitment()
+    const hashedLeaf = crypto.hashLeftRight(commitment, userStateRoot)
+    GSTree.insert(hashedLeaf)
+    const GSTreeProof = GSTree.createProof(0) // if there is only one GST leaf, the index is 0
+    const GSTreeRoot = GSTree.root
+
+    const circuitInputs = {
+        epoch,
+        identity_nullifier: id.identityNullifier,
+        identity_trapdoor: id.trapdoor,
+        user_tree_root: userStateRoot,
+        GST_path_index: GSTreeProof.pathIndices,
+        GST_path_elements: GSTreeProof.siblings,
+        attester_id: attesterId,
+        pos_rep: reputationRecords[attesterId].posRep,
+        neg_rep: reputationRecords[attesterId].negRep,
+        graffiti: reputationRecords[attesterId].graffiti,
+        sign_up: reputationRecords[attesterId].signUp,
+        UST_path_elements: USTPathElements,
+    }
```

Comment:
Sorry I missed this message
So the `graffiti_pre_image` signal is missing in the `genCircuitInput`
and the graffiti is `hash(graffitiPreImage)`
so you should set a value to graffiti preimage e.g.`graffitiPreimage = 0`
and in `reputationRecords` you should use
```ts
        const reputationRecords = {
            [attesterId]: new Reputation(
                BigInt(10),
                BigInt(0),
                hash(graffitiPreimage),
                BigInt(1)
            ),
        }
```

and also add 
```ts
graffiti_pre_image: graffitiPreimage
```
in circuit inputs